Add filters to allow themes and plugins to pass HTML attributes to different Nav Walker output

davidwebca opened the pull request:
Currently, Nav Walkers allow custom HTML attributes to be passed to the link (<a>) element only. There are many instances, like when creating a menu with a declarative JavaScript library, where it would be ideal to be able to pass down HTML attributes to any of the built element. UL and LI are currently excluded of this.

Trac ticket: https://core.trac.wordpress.org/ticket/57140

davidwebca committed:
Add filters to allow themes and plugins to pass HTML attributes to the default Nav Walker's output

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		/**
+		 * Filters the HTML attributes applied to a menu item's ul element.
+		 *
+		 * @since 6.1.2
```
```suggestion
		 * @since 6.2.0
```
The earliest that this new filter could be merged to Core is WordPress 6.2.0.

🔢 Applies elsewhere in the PR.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+		foreach ( $atts as $attr => $value ) {
+			if ( is_scalar( $value ) && '' !== $value && false !== $value ) {
+				$value       = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
+				$attributes .= ' ' . $attr . '="' . $value . '"';
```
```suggestion
				$attributes .= ' ' . esc_attr( $attr ) . '="' . $value . '"';
```
Attribute names will also need to be escaped here.

davidwebca replied:
Hi. I don't think this is applicable here. Attributes values and names have different possible characters so the escaping mechanism should be different. WordPress doesn't have an attribute name escaping mecanism yet so we should avoid adding it here since [it wasn't there before](https://github.com/WordPress/wordpress-develop/blob/d2f420493d87465c323190892d4e7a6e4c6070a7/src/wp-includes/class-walker-nav-menu.php#L209-L215). It could very well introduce breaking change to millions of users. I'll add some links down below to document previous conversations that happened about this in the last 10 years on WP core.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+	protected function build_atts( $atts = array() ) {
+		$attributes = '';
+		foreach ( $atts as $attr => $value ) {
+			if ( is_scalar( $value ) && '' !== $value && false !== $value ) {
```
```suggestion
			if ( false !== $value && is_scalar( $value ) && '' !== trim( $value ) ) {
```
1. Slight re-ordering to skip early if `$value` is `false`.
2. `trim( $value )` to prevent values with only blank space. - If this isn't preferred here, let me know and include an example of an acceptable empty value to help clarify this for me.

davidwebca replied:
I agree on the reordering, but blank values are allowed in HTML values so trimming shouldn't be added here. The main thing that I would like to further address in a future patch would be to allow empty html attributes (ex.: "disabled" without any value or <span aria-hidden>Something</span>) This would involve a bit of refactoring. I'll resolve this for now.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+		foreach ( $atts as $attr => $value ) {
+			if ( is_scalar( $value ) && '' !== $value && false !== $value ) {
+				$value       = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
+				$attributes .= ' ' . $attr . '="' . $value . '"';
```
Curious about `(bool) true` values here.

`(bool) false` are skipped, but `(bool) true` will be accepted. That is, `$attr = true` will become `attr="1"`.

Is this the intended/desired behaviour, or should `(bool) true` be treated as a boolean attribute and only `esc_attr( $attr )` be appended to `$attributes`?

davidwebca replied:
I mainly replicated what was already in use in this file to avoid creating breaking changes as much as possible. Any scalar value can be easily transformed into a string by PHP, some people might already be relying in their code on ```(bool) true``` becoming (string) "1" so I wouldn't want to step into that for now.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
-		$class_names = $class_names ? ' class="' . esc_attr( $class_names ) . '"' : '';
 
-		$output .= "{$n}{$indent}<ul$class_names>{$n}";
+		$atts 		   = array();
```
```suggestion
		$atts          = array();
```
Should fix this PHPCS failure.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		 * @param stdClass $args      An object of `wp_nav_menu()` arguments.
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$atts 		= apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
```
```suggestion
		$atts        = apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
```
Should fix this PHPCS failure.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		$atts['class'] = ! empty( $class_names ) ? $class_names : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu item's ul element.
```
```suggestion
		 * Filters the HTML attributes applied to a menu list element.
```

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		 * @param array $atts {
+		 *     The HTML attributes applied to the `<ul>` element, empty strings are ignored.
+		 *
+		 *     @type string $class    HTML CSS class attribute
```
```suggestion
		 *     @type string $class    HTML CSS class attribute.
```

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+		$li_atts['id']    = ! empty( $id ) ? $id : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu item's li element.
```
```suggestion
		 * Filters the HTML attributes applied to a menu's list item element.
```

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+		 *     The HTML attributes applied to the menu item's `<li>` element, empty strings are ignored.
+		 *
+		 *     @type string $class        HTML CSS class attribute.
+		 *     @type string $id           ID html attribute.
```
```suggestion
		 *     @type string $id           HTML id attribute.
```

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$atts 		= apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
+		$attributes = $this->build_atts($atts);
```
```suggestion
		$attributes = $this->build_atts( $atts );
```
Should fix these PHPCS failures.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
-		$id = $id ? ' id="' . esc_attr( $id ) . '"' : '';
 
-		$output .= $indent . '<li' . $id . $class_names . '>';
+		$li_atts 		  = array();
```
```suggestion
		$li_atts          = array();
```
Should fix this PHPCS failure.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$li_atts 	   = apply_filters( 'nav_menu_item_attributes', $li_atts, $menu_item, $args, $depth );
+		$li_attributes = $this->build_atts($li_atts);
```
```suggestion
		$li_attributes = $this->build_atts( $li_atts );
```
Should fix these PHPCS failures.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -204,15 +242,8 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
-				$attributes .= ' ' . $attr . '="' . $value . '"';
-			}
-		}
+		$atts 		= apply_filters( 'nav_menu_link_attributes', $atts, $menu_item, $args, $depth );
```
```suggestion
		$atts       = apply_filters( 'nav_menu_link_attributes', $atts, $menu_item, $args, $depth );
```
Should fix this PHPCS failure.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -204,15 +242,8 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
-			}
-		}
+		$atts 		= apply_filters( 'nav_menu_link_attributes', $atts, $menu_item, $args, $depth );
+		$attributes = $this->build_atts($atts);
```
```suggestion
		$attributes = $this->build_atts( $atts );
```
Should fix this PHPCS failures.

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
 	}
 
+	/**
+	 * Builds an HTML attributes string based on an array of key value pairs
+	 * Empty values are ignored
```
```suggestion
	 * Builds a string of HTML attributes from an array of key/value pairs.
	 * Empty values are ignored.
```

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+	 * Builds an HTML attributes string based on an array of key value pairs
+	 * Empty values are ignored
+	 *
+	 * @since 6.1.2
```
```suggestion
	 * @since 6.2.0
```

costdev commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+	 *
+	 * @since 6.1.2
+	 *
+	 * @param      array   $atts   The atts
+	 * @return     string  HTML attributes string
```
```suggestion
	 * @param array $atts Optional. An array of HTML attribute key/value pairs. Default empty array.
	 * @return string  A string of HTML attributes.
```

costdev reviewed (changes requested):
Hi @davidwebca, thanks for the PR!

I've left some suggestions to improve security and documentation, ask a question about boolean attributes, and help resolve the PHPCS failures.

Alternatively, for the PHPCS failures, you could run this in your terminal to autofix them:
```sh
./vendor/bin/phpcbf src/wp-includes/class-walker-nav-menu.php
```

davidwebca committed:
Fix PHPCS and target version

davidwebca committed:
Reorder conditions to fail early on strict false boolean

davidwebca committed:
Typos and grammar in docblocks

davidwebca commented:
Thanks @costdev for your pointers, very helpful. 

As for the HTML attributes building function, here's my train of thought: 

At first, I started using the [code that was already there](https://github.com/WordPress/wordpress-develop/blob/d2f420493d87465c323190892d4e7a6e4c6070a7/src/wp-includes/class-walker-nav-menu.php#L209-L215) with the assumptions that it had already been approved by core and had the least risk of causing breaking changes for end users. I had replicated the code multiple times and it started feeling silly so I extracted it into its separated protected function in the Nav Walker's class. But, then, my research lead me to previous efforts regarding this: 

- A general HTML attribute building function just like the one I extracted that could be added to formatting.php - https://core.trac.wordpress.org/ticket/23236. Opened 10 years ago and closed 9 years ago for lack of interest.
- A general HTML attribute name escape function with an interesting conversation about the differences in allowed characters to escape or not in attribute names versus values. https://core.trac.wordpress.org/ticket/43010

I think it's safe to assume that the path to have this included as fast as possible without introducing breaking changes in Nav Walkers would be to continue with this patch without involving too many different conversations at once, but once that's done, we could start looking at reviving those two conversations, especially since the second one already has a patch and is just sitting there without activity. It would be amazing to have a general HTML attributes builder functions that every classes could refactor to use and eventually debloat + secure a lot of code.

costdev commented:
@davidwebca Thanks for the detailed response!

I agree with you that given the following:
1. The existing code doesn't escape attribute names.
2. Extenders may be using spec-valid attribute names that `esc_attr()` would break.
3. Core doesn't have a spec-valid escape for attribute names.

This PR shouldn't escape the attribute names, doesn't introduce an additional security issue here, and escaping these attribute names should be part of wider discussion about the introduction of a spec-valid escape for attribute names.

Had this been newly introduced code, the introduction of a spec-valid escape in future would be an _enhancement_, not a regression. However, using `esc_attr()` in this PR would be a regression.

------

Regarding test failures, note the previous output:
```php
$output .= $indent . '<li' . $id . $class_names . '>';
```

Reversing lines 192 and 193 in the PR shows all tests passing with:
```sh
./vendor/bin/phpunit --filter '/Tests_Menu_Walker_Nav_Menu|Tests_Menu_wpNavMenu/'
```

peterwilsoncc commented:
> This PR shouldn't escape the attribute names, doesn't introduce an additional security issue here, and escaping these attribute names should be part of wider discussion about the introduction of a spec-valid escape for attribute names.

Just a quick note that WordPress prevents double escaping, so `esc_attr()` should be fine in this context. From `wp shell` to demonstrate:

```
wp> $attr = 'Pens & Pencils';
=> string(14) "Pens & Pencils"
wp> $attr = esc_attr( $attr );
=> string(18) "Pens &amp; Pencils"
wp> $attr = esc_attr( $attr );
=> string(18) "Pens &amp; Pencils"
```

costdev commented:
@peterwilsoncc So you don't anticipate any issues caused by introducing `esc_attr()` for attribute names in this PR?

If so, should this change be made in this PR, or a follow-up ticket to note the change?

peterwilsoncc commented:
@costdev No, I don't. I'm happy for the change to remain in this PR, the data is currently escaped and this PR is simply moving around where that takes place. Feel free to correct me if I am misunderstanding, I've only done a quick scan of the PR.

costdev commented:
Summary of discussion with @peterwilsoncc in Slack:
- [The existing code](https://github.com/WordPress/wordpress-develop/blob/d2f420493d87465c323190892d4e7a6e4c6070a7/src/wp-includes/class-walker-nav-menu.php#L209-L215) doesn't escape attribute names, only values.
- Introducing `esc_attr()` to attribute names here could break CSS styling.
- Further consideration is needed before escaping attribute names in already-existing code.

In short: The PR can continue without introducing `esc_attr()` for attribute names.

davidwebca committed:
Reverse id and class order to pass tests

davidwebca commented:
I've made the swap suggested by @costdev and tests seem to be okay.

davidwebca commented:
Hi @costdev! Just wanted to check on this. Is there anything else I can do to help get the ball rolling or is this just waiting to be pulled in?

davidwebca committed:
Adding tests by costdev - thanks

davidwebca committed:
Revert "Adding tests by costdev - thanks"

This reverts commit 4c5e46794d5ecbe3d49dcd3aab55a59ebd0b292e.

davidwebca committed:
Merge branch 'WordPress:trunk' into trunk

davidwebca committed:
Adding tests by costdev - thanks

davidwebca commented:
Tests are added. I didn't read them through but since the patch was given by @costdev I have full confidence it's ok. Sorry for the multiple commits, there were some conflicts when applying the patch.

costdev commented on tests/phpunit/tests/menu/walker-nav-menu.php:
```diff
@@ -365,4 +365,117 @@ public function test_walker_nav_menu_start_el_should_add_rel_privacy_policy_when
+	 *
+	 * @ticket 57278
+	 *
+	 * @covers Walker_Nav_Menu::start_el()
```
```suggestion
	 * @covers Walker_Nav_Menu::start_el
```

costdev commented on tests/phpunit/tests/menu/walker-nav-menu.php:
```diff
@@ -365,4 +365,117 @@ public function test_walker_nav_menu_start_el_should_add_rel_privacy_policy_when
+	 *
+	 * @ticket 57278
+	 *
+	 * @covers Walker_Nav_Menu::build_atts()
```
```suggestion
	 * @covers Walker_Nav_Menu::build_atts
```

costdev commented on tests/phpunit/tests/menu/walker-nav-menu.php:
```diff
@@ -365,4 +365,117 @@ public function test_walker_nav_menu_start_el_should_add_rel_privacy_policy_when
+	 *
+	 * @ticket 57278
+	 *
+	 * @covers Walker_Nav_Menu::start_lvl()
```
```suggestion
	 * @covers Walker_Nav_Menu::start_lvl
```

costdev reviewed (changes requested):
Thanks for adding the tests @davidwebca!

I accidentally included `()` in the `@covers` annotations. I've left the suggestions to remove these in this review 🙂

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		/**
+		 * Filters the HTML attributes applied to a menu list element.
+		 *
+		 * @since 6.2.0
```
```suggestion
		 * @since 6.3.0
```

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+		/**
+		 * Filters the HTML attributes applied to a menu's list item element.
+		 *
+		 * @since 6.2.0
```
```suggestion
		 * @since 6.3.0
```

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+	 * Builds a string of HTML attributes from an array of key/value pairs.
+	 * Empty values are ignored.
+	 *
+	 * @since 6.2.0
```
```suggestion
	 * @since 6.3.0
```

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
+		$atts       = apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
+		$attributes = $this->build_atts( $atts );
+
+		$output .= "{$n}{$indent}<ul$attributes>{$n}";
```
Nit: for consistency within the string.

```suggestion
		$output .= "{$n}{$indent}<ul{$attributes}>{$n}";
```

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+	 * @return string A string of HTML attributes.
+	 */
+	protected function build_atts( $atts = array() ) {
+		$attributes = '';
```
Having `$atts`, `$attributes` and `$attr` all within one function is a little unclear, let's change the outcome to something clearer. (I know this was an existing issue but it's good to improve while moving the code.)

```suggestion
		$attribute_string = '';
```

costdev replied:
By the same token I'd consider renaming the method to `build_attributes()` in the context of "you don't build bricks/$atts, you build houses/$attributes". Besides, `build_attributes` requires no expansion, so it's cleaner.

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+		$attributes = '';
+		foreach ( $atts as $attr => $value ) {
+			if ( false !== $value && '' !== $value && is_scalar( $value ) ) {
+				$value       = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
+				$attributes .= ' ' . $attr . '="' . $value . '"';
```
```suggestion
				$value             = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
				$attribute_string .= ' ' . $attr . '="' . $value . '"';
```

peterwilsoncc commented on src/wp-includes/class-walker-nav-menu.php:
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
+				$attributes .= ' ' . $attr . '="' . $value . '"';
+			}
+		}
+		return $attributes;
```
```suggestion
		return $attribute_string;
```

peterwilsoncc reviewed (commented):
I've added a few notes inline but they're all minor.

audrasjb committed:
fix since mention

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

audrasjb committed:
fix since mention

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

audrasjb committed:
fix since mention

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

davidwebca committed:
Update tests/phpunit/tests/menu/walker-nav-menu.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

davidwebca committed:
Update tests/phpunit/tests/menu/walker-nav-menu.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

davidwebca committed:
Update src/wp-includes/class-walker-nav-menu.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

davidwebca committed:
Update tests/phpunit/tests/menu/walker-nav-menu.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

davidwebca committed:
Update src/wp-includes/class-walker-nav-menu.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

davidwebca committed:
Update src/wp-includes/class-walker-nav-menu.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

davidwebca committed:
Update src/wp-includes/class-walker-nav-menu.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

davidwebca commented:
Thanks for all the suggestions. I've added them now.

peterwilsoncc commented:
Committed in [r56067](https://core.trac.wordpress.org/changeset/56067)
